packages/taiko-client/proposer/proposer.go

```diff
@@ -518,3 +610,118 @@ func (p *Proposer) sendTx(ctx context.Context, txCandidate *txmgr.TxCandidate) e
 func (p *Proposer) Name() string {
 	return "proposer"
 }
+
+// isProfitable checks if a transaction list is profitable to propose
+// Profitability is determined by comparing the revenue from transaction fees
+// to the costs of proposing and proving the block.
+func (p *Proposer) isProfitable(txLists []types.Transactions, proposingCosts *big.Int) (bool, error) {
+	totalTransactionFees, err := p.calculateTotalL2TransactionsFees(txLists)
+	if err != nil {
+		return false, err
+	}
+
+	costs, err := p.estimateTotalCosts(proposingCosts)
+	if err != nil {
+		return false, err
+	}
+
+	log.Debug("isProfitable", "total L2 fees", totalTransactionFees, "total L1 costs", costs)
+
+	return totalTransactionFees.Cmp(costs) > 0, nil
+}
+
+func (p *Proposer) calculateTotalL2TransactionsFees(txLists []types.Transactions) (*big.Int, error) {
+	totalGasConsumed := new(big.Int)
+
+	for _, txs := range txLists {
+		for _, tx := range txs {
+			baseFee := get75PercentOf(tx.GasFeeCap())
+			multiplier := new(big.Int).Add(tx.GasTipCap(), baseFee)
+			gasConsumed := new(big.Int).Mul(multiplier, baseFee)
+			totalGasConsumed.Add(totalGasConsumed, gasConsumed)
+		}
+	}
+
+	return totalGasConsumed, nil
+}
+
+func get75PercentOf(num *big.Int) *big.Int {
+	// First multiply by 3 to get 75% (as 3/4 = 75%)
+	result := new(big.Int).Mul(num, big.NewInt(3))
+	// Then divide by 4
+	return new(big.Int).Div(result, big.NewInt(4))
+}
+
+func (p *Proposer) getTransactionCost(txCandidate *txmgr.TxCandidate) (*big.Int, error) {
+	// Get the current L1 gas price
+	gasPrice, err := p.rpc.L1.SuggestGasPrice(p.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getTransactionCost: failed to get gas price: %w", err)
+	}
+
+	estimatedGasUsage, err := p.rpc.L1.EstimateGas(p.ctx, ethereum.CallMsg{
```

Comment:
@davidtaikocha I remembered that `EstimateGas` for blob tx is not very accurate. Did I misremember?